src/mlib/int128.h

```diff
@@ -0,0 +1,683 @@
+      }
+   } else if (off < 0) {
+      off = -off;
+      if (off >= 64) {
+         off -= 64;
+         uint64_t low = val.r.hi >> off;
+         return MLIB_INIT (mlib_int128) MLIB_INT128_FROM_PARTS (low, 0);
+      } else {
+         uint64_t high = val.r.hi >> off;
+         uint64_t low = val.r.lo >> off;
+         low |= val.r.hi << (64 - off);
+         return MLIB_INIT (mlib_int128) MLIB_INT128_FROM_PARTS (low, high);
+      }
+   } else {
+      return val;
+   }
+}
+
+/**
+ * @brief Bitwise logical right-shift a 128-bit integer
+ *
+ * @param val The value to modify. No "sign bit" is respected.
+ * @param off The offset to shift right. If negative, shifts left
+ * @return The result of the shift operation
+ */
+static mlib_constexpr_fn mlib_int128
+mlib_int128_rshift (mlib_int128 val, int off)
+{
+   return mlib_int128_lshift (val, -off);
+}
+
+/**
+ * @brief Bitwise-or two 128-bit integers
+ */
+static mlib_constexpr_fn mlib_int128
+mlib_int128_bitor (mlib_int128 l, mlib_int128 r)
+{
+   return MLIB_INIT (mlib_int128)
+      MLIB_INT128_FROM_PARTS (l.r.lo | r.r.lo, l.r.hi | r.r.hi);
+}
+
+// Multiply two 64bit integers to get a 128-bit result without overflow
+static mlib_constexpr_fn mlib_int128
+_mlibUnsignedMult128 (uint64_t left, uint64_t right)
+{
+   // Perform a Knuth 4.3.1M multiplication
+   uint32_t u[2] = {(uint32_t) left, (uint32_t) (left >> 32)};
+   uint32_t v[2] = {(uint32_t) right, (uint32_t) (right >> 32)};
+   uint32_t w[4] = {0};
+
+   for (int j = 0; j < 2; ++j) {
+      uint64_t t = 0;
+      for (int i = 0; i < 2; ++i) {
+         t += (uint64_t) (u[i]) * v[j] + w[i + j];
+         w[i + j] = (uint32_t) t;
+         t >>= 32;
+      }
+      w[j + 2] = (uint32_t) t;
+   }
+
+   return MLIB_INIT (mlib_int128) MLIB_INT128_FROM_PARTS (
+      ((uint64_t) w[1] << 32) | w[0], ((uint64_t) w[3] << 32) | w[2]);
+}
+
+/**
+ * @brief Multiply two mlib_int128s together. Overflow will wrap.
+ */
+static mlib_constexpr_fn mlib_int128
+mlib_int128_mul (mlib_int128 l, mlib_int128 r)
+{
+   // Multiply the low-order word
+   mlib_int128 ret = _mlibUnsignedMult128 (l.r.lo, r.r.lo);
+   // Accumulate the high-order parts:
+   ret.r.hi += l.r.lo * r.r.hi;
+   ret.r.hi += l.r.hi * r.r.lo;
+   return ret;
+}
+
+/// Get the number of leading zeros in a 64bit number.
+static mlib_constexpr_fn int
+_mlibCountLeadingZeros_u64 (uint64_t bits)
+{
+   int n = 0;
+   if (bits == 0) {
+      return 64;
+   }
+   while (!(1ull << 63 & bits)) {
+      ++n;
+      bits <<= 1;
+   }
+   return n;
+}
+
+/// Implementation of Knuth's algorithm 4.3.1 D for unsigned integer division
+static mlib_constexpr_fn void
+_mlibKnuth431D (uint32_t *const u,
+                const int ulen,
+                const uint32_t *const v,
+                const int vlen,
+                uint32_t *quotient)
+{
+   // Part D1 (normalization) is done by caller, normalized in u and v (b is 32)
```

Comment:
```suggestion
   // Part D1 (normalization) is done by caller, normalized in u and v (b is 2^32)
```

b is the radix. The digits are 32 bit unsigned integers.